File: wrappers/python/src/checkrd/hooks.py
"""Request lifecycle hooks for the Checkrd SDK.

Hooks let users run custom logic at policy evaluation time without
modifying the SDK internals. Three hook points are available:

- ``before_request`` — fires before policy evaluation. Returning ``None``
  skips evaluation entirely (pass-through). Returning the event
  (possibly modified) proceeds normally. Use this for request-level
  overrides, sampling, or cost control.
- ``on_allow`` — fires after a request is allowed by the policy engine.
  Use this for allow-side observability (metrics, structured logging).
- ``on_deny`` — fires after a request is denied (both in enforce and
  dry-run modes). Use this for alerting (Slack, PagerDuty), audit
  trails, or custom deny behavior.

Hook exceptions are caught and logged at WARNING level. A crashing hook
never takes down the user's request. This follows the Sentry and
OpenTelemetry precedent.

Usage::

    import checkrd

    def alert_on_deny(event: checkrd.CheckrdEvent) -> None:
        slack.post(f"Blocked: {event.method} {event.url} — {event.deny_reason}")

    checkrd.init(policy="policy.yaml", on_deny=alert_on_deny)
    checkrd.instrument()

Or per-client::

    client = checkrd.wrap(
        httpx.Client(),
        on_deny=lambda e: print(f"Denied: {e.url}"),
    )
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Callable, Optional


@dataclass
class CheckrdEvent:
    """Payload passed to lifecycle hooks.

    Populated progressively: ``before_request`` receives an event with
    only the request fields set (``allowed``, ``deny_reason``, etc. are
    ``None``). ``on_allow`` and ``on_deny`` receive a fully populated
    event after policy evaluation.

    Attributes:
        method: HTTP method (``GET``, ``POST``, etc.).
        url: The full request URL.
        headers: Request headers as a list of ``(name, value)`` tuples.
        body: Request body as a string, or ``None`` for empty/binary bodies.
        request_id: Unique correlation ID for this evaluation.
        allowed: ``True`` if the policy engine allowed the request,
            ``False`` if denied, ``None`` before evaluation.
        rule_name: Name of the matching rule, or ``None``.
        deny_reason: Human-readable deny reason, or ``None`` if allowed.
        suggestion: Actionable fix suggestion, or ``None``.
        dashboard_url: Deep link to the event in the dashboard, or ``None``.
    """

    method: str
    url: str
    headers: list[tuple[str, str]] = field(default_factory=list)
    body: Optional[str] = None
    request_id: str = ""
    allowed: Optional[bool] = None
    rule_name: Optional[str] = None
    deny_reason: Optional[str] = None
    suggestion: Optional[str] = None
    dashboard_url: Optional[str] = None
    #: W3C Trace Context trace-id (32 lowercase hex chars), extracted
    #: from the request's ``traceparent`` header when present so
    #: hook callers can correlate the SDK's policy decision with
    #: the user's distributed-trace span. ``None`` when the request
    #: carries no ``traceparent``. Mirrors the value the telemetry
    #: batcher stamps on its ``POST /v1/telemetry`` so a single
    #: ``trace_id`` spans agent code → policy eval → telemetry
    #: ingestion → ClickHouse.
    trace_id: Optional[str] = None


#: Callback invoked before policy evaluation. Return ``None`` to skip
#: evaluation (pass-through), or return the event to proceed.
BeforeRequestHook = Callable[[CheckrdEvent], Optional[CheckrdEvent]]

#: Callback invoked after a request is allowed by the policy engine.
OnAllowHook = Callable[[CheckrdEvent], None]

#: Callback invoked after a request is denied (both enforce and dry-run).
OnDenyHook = Callable[[CheckrdEvent], None]


#: Hint metadata passed to :data:`BeforeSendHook`. Populated by the SDK;
#: users read but don't mutate. Mirrors Sentry's ``hint`` argument: gives
#: the hook context that's not stored in the event itself (e.g., what
#: agent the event came from, whether it's a stream completion or a
#: regular request evaluation). Stable keys are documented below; we
#: reserve the right to add more in minor releases (callers should
#: ``.get(key, default)`` rather than assume keys exist).
#:
#: Documented keys:
#:   - ``agent_id`` (str): the agent emitting the event.
#:   - ``event_kind`` (str): ``"request_evaluation"`` for the policy-
#:     decision event the transport enqueues, ``"stream_completion"``
#:     for the post-stream token-usage event, etc.
BeforeSendHint = "dict[str, object]"


#: Telemetry-event mutation/drop hook. Fires once per event right
#: before it's enqueued for batched delivery. Use it to:
#:
#: - **redact fields**: ``event.pop("body_hash", None); return event``
#: - **drop events**: ``return None`` (the event never ships; no
#:   ``dropped_*`` counter increments — operator-intended drops are
#:   not failures)
#: - **transform payloads**: rewrite endpoint URLs, normalize status
#:   codes, attach static labels, etc.
#:
#: Same name and same contract as Sentry's ``before_send``: returning
#: ``None`` drops, returning the (possibly mutated) event ships it.
#: Operators migrating from Sentry recognize the shape immediately.
#:
#: Hook exceptions are caught and logged; a crashing hook drops the
#: event but never takes down the calling thread or the user's
#: request critical path.
BeforeSendHook = Callable[
    ["dict[str, object]", "dict[str, object]"], Optional["dict[str, object]"],
]

File: wrappers/javascript/src/hooks.ts
/**
 * Callback hook types mirroring `checkrd/hooks.py`. Hook implementations
 * must be pure and non-blocking — they run on the request critical path.
 */

/** Metadata passed to every hook. Headers are redacted before they land here. */
export interface CheckrdEvent {
  /** HTTP method (upper-case, e.g., "GET"). */
  method: string;
  /** Full request URL. */
  url: string;
  /** Whether the request was allowed (undefined for before_request). */
  allowed?: boolean;
  /** Deny reason, set only when allowed is false. */
  denyReason?: string;
  /** Correlation ID. */
  requestId: string;
  /** Request body, if available and under 1 MB. */
  body?: string;
  /** Request headers with auth-related values redacted. */
  headers: [string, string][];
  /**
   * W3C Trace Context trace-id (32 lowercase hex chars), extracted
   * from the request's ``traceparent`` header when present so hook
   * callers can correlate the SDK's policy decision with the user's
   * distributed-trace span. ``undefined`` when the request carries
   * no ``traceparent``. Mirrors the value the telemetry batcher
   * stamps on its ``POST /v1/telemetry`` so a single ``trace_id``
   * spans agent code → policy eval → telemetry ingestion → ClickHouse.
   */
  traceId?: string;
}

/**
 * Cross-realm Symbol used as the property key for the SDK's
 * correlation request-id on every wrapped fetch ``Response``. Callers
 * read it to tie a specific call to a telemetry event without
 * re-instrumenting the request path:
 *
 *     const response = await myFetch(url, init);
 *     const requestId = (response as Response & { [CHECKRD_REQUEST_ID]?: string })[CHECKRD_REQUEST_ID];
 *     // → paste into a Checkrd support ticket and the on-call can
 *     //   locate the exact policy evaluation row.
 *
 * Symbol.for("checkrd.request_id") makes the same Symbol resolvable
 * from Workers, dynamic imports, and other-realm code paths so a
 * caller using a different SDK instance still gets the same key.
 * Mirrors the OpenAI Node SDK's ``_request_id`` convention but uses
 * a Symbol (not a string property) to avoid colliding with vendor
 * SDK fields.
 */
export const CHECKRD_REQUEST_ID = Symbol.for("checkrd.request_id");

/**
 * Type alias for a ``Response`` carrying Checkrd's request-id. Use
 * when you need the strict type of a wrapped response without the
 * cast at the call site:
 *
 *     const response: CheckrdResponse = await myFetch(url, init);
 *     const id = response[CHECKRD_REQUEST_ID];
 */
export type CheckrdResponse = Response & {
  readonly [CHECKRD_REQUEST_ID]?: string;
};

/**
 * Extract the trace-id (32 hex chars) from a W3C ``traceparent``
 * header. Returns ``undefined`` for any malformed input rather than
 * throwing — a bad header from customer code must never break the
 * request path.
 *
 * Format: ``{version}-{trace-id}-{parent-id}-{flags}`` with
 * ``version=00`` for the current spec. We accept the version field
 * but drop traces from any other version (forward-compatible
 * extraction can't reliably parse unknown versions).
 */
export function extractTraceId(
  headers: [string, string][] | Headers,
): string | undefined {
  let raw: string | null | undefined;
  if (headers instanceof Headers) {
    raw = headers.get("traceparent");
  } else {
    for (const [k, v] of headers) {
      if (k.toLowerCase() === "traceparent") {
        raw = v;
        break;
      }
    }
  }
  if (raw === null || raw === undefined || typeof raw !== "string") return undefined;
  const parts = raw.split("-");
  if (parts.length !== 4) return undefined;
  const [version, traceId] = parts as [string, string, string, string];
  if (version !== "00") return undefined;
  if (traceId.length !== 32 || !/^[0-9a-f]{32}$/.test(traceId)) return undefined;
  // All-zero trace-id is invalid per W3C spec.
  if (traceId === "0".repeat(32)) return undefined;
  return traceId;
}

/** Fires after policy evaluation if the request was allowed. Must return quickly. */
export type OnAllowHook = (event: CheckrdEvent) => void;

/**
 * Fires after policy evaluation if the request was denied. Must return quickly.
 * Enforcement is already decided by this point; the hook cannot reverse it.
 */
export type OnDenyHook = (event: CheckrdEvent) => void;

/**
 * Fires before policy evaluation. Return `false` to short-circuit the
 * request (it will not be evaluated or forwarded); return `true` or
 * `undefined` to proceed normally.
 */
export type BeforeRequestHook = (event: CheckrdEvent) => boolean | undefined;
